app/utility/base_planning_svc.py

```diff
@@ -113,6 +117,38 @@ async def obfuscate_commands(self, agent, obfuscator, links):
 
     """ PRIVATE """
 
+    @staticmethod
+    def _list_historic_duplicate_singletons(operation):
+        """
+        Generate a list of successfully run singleton abilities for a given operation
+        :param operation: Operation to scan
+        :return: List of command hashes for succeeded singleton abilities
+        """
+        singleton = [k for k in operation.chain if k.status == k.states['SUCCESS'] and k.ability.singleton]
+        return [(x.command_hash if x.command_hash else x.command) for x in singleton if x]
+
+    @staticmethod
+    def _cross_check_agents_for_duplicate_singletons(agent_links):
```

Comment:
rename to 'remove_links_...' for consistency

Reply:
Can do.